### PR TITLE
Laboratory work #3, Vsevolod Papin - 23FPL2

### PR DESCRIPTION
**Pull request by AllKnowingGuy**
I couldn't add a new pull request, so I just renamed the last one

### CODE REVIEW

File: lab_1_classify_by_unigrams/main.py [hunk 1]
```diff
@@ -25,11 +24,9 @@ def tokenize(text: str) -> list[str] | None:
     """
     if not isinstance(text, str):
         return None
-    tokenized_list = []
-    for token in text.lower():
-        if token.isalpha():
-            tokenized_list.append(token)
-    return tokenized_list
+
```
Comment (Vasilisa-Blyudova) [inline lab_1_classify_by_unigrams/main.py:27]: revert all changes in lab1

**Review by Vasilisa-Blyudova — COMMENTED** [review]
Unfortunately, I don’t see any progress in laboratory work 2 yet. I advise you to start implementing

File: lab_1_classify_by_unigrams/main.py (hunk 1 above)
Comment (AllKnowingGuy) [reply lab_1_classify_by_unigrams/main.py]: Done

**Comment by AllKnowingGuy** [comment]
Will fix MyPy issues during the afternoon

File: lab_2_retrieval_w_bm25/start.py [hunk 2]
```diff
@@ -3,6 +3,8 @@
 """
 # pylint:disable=too-many-locals, unused-argument, unused-variable, too-many-branches, too-many-statements, duplicate-code
 
+import lab_2_retrieval_w_bm25.main as funcs
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/start.py:6]: import required functions from main

File: lab_2_retrieval_w_bm25/main.py [hunk 3]
```diff
@@ -18,6 +20,17 @@ def tokenize(text: str) -> list[str] | None:
 
     In case of corrupt input arguments, None is returned.
     """
+    if not isinstance(text, str):
+        return None
+
+    symbol_list = list(text)
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:26]: don't creat list, try to use replace string method

File: lab_2_retrieval_w_bm25/main.py [hunk 4]
```diff
@@ -33,6 +46,20 @@ def remove_stopwords(tokens: list[str], stopwords: list[str]) -> list[str] | Non
...
+            len(tokens) > 0 and
+            all(isinstance(token, str) for token in tokens) and
+            isinstance(stopwords, list) and
+            len(stopwords) > 0 and
+            all(isinstance(word, str) for word in stopwords)):
+        return None
+
+    new_tokens = tokens[:]
+    for token in tokens:
+        if token in stopwords:
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:59]: use list comprehension

File: lab_2_retrieval_w_bm25/main.py [hunk 5]
```diff
@@ -92,6 +163,22 @@ def calculate_tf_idf(tf: dict[str, float], idf: dict[str, float]) -> dict[str, f
...
+    if not (isinstance(tf, dict) and
+            len(tf) > 0 and
+            all(isinstance(key, str) and isinstance(tf[key], float) for key in tf) and
+            isinstance(idf, dict) and
+            len(idf) > 0 and
+            all(isinstance(key, str) and isinstance(idf[key], float) for key in idf)):
+        return None
+
+    tf_idf = {}
+    for key in tf:
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:175]: do we have different keys for tf and idf dictionaries?

File: lab_2_retrieval_w_bm25/main.py [hunk 6]
```diff
@@ -138,6 +256,29 @@ def rank_documents(
...
+        return None
+
+    tokenized_query = tokenize(query)
+    query_to_compare = remove_stopwords(tokenized_query, stopwords)
+    if not isinstance(query_to_compare, list):
+        return None
+
+    doc_sums = []
+    for doc_num in range(len(indexes)):
+        s = 0
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:275]: rename variable s

**Review by Vasilisa-Blyudova — COMMENTED** [review]
Very good work! Implement it to the 10 score!

File: lab_2_retrieval_w_bm25/main.py (hunk 5 above)
Comment (AllKnowingGuy) [reply lab_2_retrieval_w_bm25/main.py:175]: This is a check: IDF should have all the keys of TF, and if it doesn't, None is returned.

File: lab_2_retrieval_w_bm25/main.py (hunk 3 above)
Comment (AllKnowingGuy) [reply lab_2_retrieval_w_bm25/main.py:26]: Done

File: lab_2_retrieval_w_bm25/start.py (hunk 2 above)
Comment (AllKnowingGuy) [reply lab_2_retrieval_w_bm25/start.py:6]: Done

File: lab_2_retrieval_w_bm25/main.py (hunk 4 above)
Comment (AllKnowingGuy) [reply lab_2_retrieval_w_bm25/main.py:59]: Done

File: lab_2_retrieval_w_bm25/main.py (hunk 5 above)
Comment (AllKnowingGuy) [reply lab_2_retrieval_w_bm25/main.py:175]: Example: the function accidentally takes TF of a text that wasn't used in calculating IDF

File: lab_2_retrieval_w_bm25/main.py (hunk 6 above)
Comment (AllKnowingGuy) [reply lab_2_retrieval_w_bm25/main.py:275]: Done

File: lab_2_retrieval_w_bm25/main.py [hunk 7]
```diff
@@ -138,6 +252,31 @@ def rank_documents(
...
+    query_to_compare = remove_stopwords(tokenized_query, stopwords)
+    if not isinstance(query_to_compare, list):
+        return None
+
+    doc_sums = []
+    for doc_num, doc in enumerate(indexes):
+        values_sum = 0.0
+        for token in query_to_compare:
+            if token in doc:
+                values_sum += doc[token]
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:276]: can we use `sum()`?

File: lab_2_retrieval_w_bm25/main.py [hunk 8]
```diff
@@ -192,6 +366,12 @@ def load_index(file_path: str) -> list[dict[str, float]] | None:
 
     In case of corrupt input arguments, None is returned.
     """
+    if isinstance(file_path, str) and len(file_path) > 0:
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:369]: return `None` if not isinstance

File: lab_2_retrieval_w_bm25/main.py [hunk 9]
```diff
@@ -178,6 +348,10 @@ def save_index(index: list[dict[str, float]], file_path: str) -> None:
         index (list[dict[str, float]]): The index to save.
         file_path (str): The path to the file where the index will be saved.
     """
+    if (isinstance(index, list) and all(isinstance(doc, dict) for doc in index) and
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:351]: return `None` if not isinstance

**Review by Vasilisa-Blyudova — COMMENTED** [review]
Great work! Now you need to fix the comments. Good luck!

File: lab_2_retrieval_w_bm25/main.py [hunk 10]
```diff
@@ -18,6 +21,16 @@ def tokenize(text: str) -> list[str] | None:
...
     In case of corrupt input arguments, None is returned.
     """
+    if not isinstance(text, str):
+        return None
+
+    for symbol in text:
+        if not (symbol.isalpha() or
+                symbol == " " or
+                symbol == "\n"):
+            text = text.replace(symbol, " ")
```
[the remark this answers is not in the document]
Comment (AllKnowingGuy) [reply lab_2_retrieval_w_bm25/main.py:31]: Done

File: lab_2_retrieval_w_bm25/main.py [hunk 11]
```diff
@@ -47,6 +70,19 @@ def build_vocabulary(documents: list[list[str]]) -> list[str] | None:
...
+    if not (isinstance(documents, list) and
+            len(documents) > 0 and
+            all(isinstance(doc, list) for doc in documents)):
+        return None
+
+    unique_words = set()
+    for doc in documents:
+        if all(isinstance(token, str) for token in doc):
+            unique_words = unique_words | set(doc)
+        else:
```
[the remark this answers is not in the document]
Comment (AllKnowingGuy) [reply lab_2_retrieval_w_bm25/main.py:82]: Yes, it excludes filling the vocabulary list with numbers and other non-words

[the remark this answers is not in the document]
Comment (AllKnowingGuy) [reply lab_2_retrieval_w_bm25/main.py:82]: Plus, the documents should only contain strings, so if the function finds something else, they are considered corrupt arguments and cause the return of None

File: lab_2_retrieval_w_bm25/main.py (hunk 7 above)
Comment (AllKnowingGuy) [reply lab_2_retrieval_w_bm25/main.py:276]: Probably yes, if MyPy is fine with that

File: lab_2_retrieval_w_bm25/main.py (hunk 8 above)
Comment (AllKnowingGuy) [reply lab_2_retrieval_w_bm25/main.py:369]: Done

File: lab_2_retrieval_w_bm25/main.py (hunk 9 above)
Comment (AllKnowingGuy) [reply lab_2_retrieval_w_bm25/main.py:351]: Done

File: lab_2_retrieval_w_bm25/main.py (hunk 7 above)
Comment (AllKnowingGuy) [reply lab_2_retrieval_w_bm25/main.py]: Yes, we can

Comment (Vasilisa-Blyudova) [reply lab_2_retrieval_w_bm25/main.py]: Great!

**Pull request closed without merging**